### PR TITLE
Enhance AWS APM metrics mapping implementation

### PR DESCRIPTION
**Pull request by mxiamxia**
**Description:**
Enhance AWS APM metrics mapping implementation to resolve a few of the following "Unknown" metric attributes use cases.

- if span name used in ingress operation is not valid with values of 'null', 'UnknownOperation' and 'HttpMethod value` for HTTP request use cases. We will try to extract the ingress operation from `http.method` and `http.target` attributes.
- if the remote service attribute can not be directly derived from the attributes such `rpc.name`, `db.system`. We will try to fill it with the `net.peer.name` and `net.sock.peer.addr` attribute values and have Collector to do the further resolution on the populated values. 
- if the remote operation attribute is `Unknown`, We will try to extract the remote operation from  `http.method` and `http.url` attributes.
- Add a new `aws.remote.target` attribute for supporting the following 4 AWS resources when invoking it with Client Spans  
   - AWS S3 Bucket Name
   - AWS Queue Name
   - AWS Stream Name
   - AWS DynamoDB table name

**Testing:**
unit test and adhoc

**Documentation:**
N/A

**Outstanding items:**
N/A

### CODE REVIEW

File: aws-xray/src/main/java/io/opentelemetry/contrib/awsxray/AwsMetricAttributeGenerator.java
Comment (bjrara) [inline aws-xray/src/main/java/io/opentelemetry/contrib/awsxray/AwsMetricAttributeGenerator.java:237]: It's hard to tell the difference between `generateRemoteService/Operation` and `getRemoteService/Operation`, only that `generate*` does some specific to HTTP/NET spans which makes me feel the method name is ambiguous.

I'm thinking if it will be better to assign `UNKNOWN*` as the initial values to `remoteService` and `remoteOperation`, and only do HTTP/NET handling using function like `getRemote*FromNETSpan` in the `else` block.

Comment (bjrara) [reply aws-xray/src/main/java/io/opentelemetry/contrib/awsxray/AwsMetricAttributeGenerator.java]: Thanks! It's much clear now.

**Review by bjrara — APPROVED** [review]
LGTM

**Comment by mxiamxia** [comment]
Hi @trask , could you please take a look this PR as code approver when you get the chance? Thanks!

**Review by bjrara — APPROVED** [review]

**Comment by wangzlei** [comment]
LGTM!

**Review by wangzlei — APPROVED** [review]

File: aws-xray/src/main/java/io/opentelemetry/contrib/awsxray/AwsMetricAttributeGenerator.java
Comment (thpierce) [inline aws-xray/src/main/java/io/opentelemetry/contrib/awsxray/AwsMetricAttributeGenerator.java:231]: I thought this would be: `{messaging.destination.name} {messaging.operation}`

File: aws-xray/src/main/java/io/opentelemetry/contrib/awsxray/AwsMetricAttributeGenerator.java
Comment (thpierce) [inline aws-xray/src/main/java/io/opentelemetry/contrib/awsxray/AwsMetricAttributeGenerator.java:242]: What is un-directive?

File: aws-xray/src/main/java/io/opentelemetry/contrib/awsxray/AwsMetricAttributeGenerator.java
Comment (thpierce) [inline aws-xray/src/main/java/io/opentelemetry/contrib/awsxray/AwsMetricAttributeGenerator.java:135]: Please update comment, behaviour has substantially changed. Please review other comments in this code and update accordingly.

File: aws-xray/src/main/java/io/opentelemetry/contrib/awsxray/AwsMetricAttributeGenerator.java
Comment (thpierce) [inline aws-xray/src/main/java/io/opentelemetry/contrib/awsxray/AwsMetricAttributeGenerator.java:268]: if `HTTP_METHOD` is not found, are we sure we want to keep just the first part of the target?

Also this code is confusing because in `isValidOperation`, we are saying that HTTP_METHOD alone is invalid, but here we consider it valid if target is null. A cleaner solution would be to update the if block on line 158 to attempt to append target if only HTTP_METHOD is being used as span name.

File: aws-xray/src/main/java/io/opentelemetry/contrib/awsxray/AwsMetricAttributeGenerator.java
Comment (thpierce) [inline aws-xray/src/main/java/io/opentelemetry/contrib/awsxray/AwsMetricAttributeGenerator.java:302]: Not blocking: This doesn't seem appropriate for this method - I would probably keep it in setRemoteServiceAndOperation. Ditto to generateRemoteService

File: aws-xray/src/main/java/io/opentelemetry/contrib/awsxray/AwsMetricAttributeGenerator.java
Comment (thpierce) [inline aws-xray/src/main/java/io/opentelemetry/contrib/awsxray/AwsMetricAttributeGenerator.java:313]: Should we do null as well?

File: aws-xray/src/test/java/io/opentelemetry/contrib/awsxray/AwsMetricAttributeGeneratorTest.java
Comment (thpierce) [inline aws-xray/src/test/java/io/opentelemetry/contrib/awsxray/AwsMetricAttributeGeneratorTest.java:6]: Did not have time to review tests, will rely on reviews from others to cover these.

File: aws-xray/src/main/java/io/opentelemetry/contrib/awsxray/AwsMetricAttributeGenerator.java
Comment (thpierce) [inline aws-xray/src/main/java/io/opentelemetry/contrib/awsxray/AwsMetricAttributeGenerator.java:214]: I thought we said that remote service/remote operation would be done independently (e.g. could have rpc service and db operation)

File: aws-xray/src/main/java/io/opentelemetry/contrib/awsxray/AwsAttributeKeys.java
Comment (carolabadeer) [reply aws-xray/src/main/java/io/opentelemetry/contrib/awsxray/AwsAttributeKeys.java:30]: +1. I was going to recommend adding a TODO note here to update these attributes to match the semantic attributes, but I prefer @thpierce suggestion

Comment (mxiamxia) [reply aws-xray/src/main/java/io/opentelemetry/contrib/awsxray/AwsAttributeKeys.java:30]: The listed AWS resource attributes are defined in [opentelemetry-java-instrumentation#AwsExperimentalAttributes ](https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/0070148b3424db5b11be257b1e086cf8ba59ed0b/instrumentation/aws-sdk/aws-sdk-1.11/library/src/main/java/io/opentelemetry/instrumentation/awssdk/v1_11/AwsExperimentalAttributes.java#L15) which is not visible to SDK. I think the long term solution is to move these resource name definitions into [semconv#SemanticAttributes](https://github.com/open-telemetry/opentelemetry-java/blob/main/semconv/src/main/java/io/opentelemetry/semconv/trace/attributes/SemanticAttributes.java). 

I think we have to define it again here before we can add it into semconv package.

File: aws-xray/src/main/java/io/opentelemetry/contrib/awsxray/AwsMetricAttributeGenerator.java
Comment (mxiamxia) [reply aws-xray/src/main/java/io/opentelemetry/contrib/awsxray/AwsMetricAttributeGenerator.java:231]: thanks for calling it out. I am thinking moving `{messaging.destination.name}` to `aws.remote.target` attribute later and keep `remote.operation` with `messaging.operation`, what do you think?

File: aws-xray/src/main/java/io/opentelemetry/contrib/awsxray/AwsMetricAttributeGenerator.java
Comment (mxiamxia) [reply aws-xray/src/main/java/io/opentelemetry/contrib/awsxray/AwsMetricAttributeGenerator.java:242]: removed it.

File: aws-xray/src/main/java/io/opentelemetry/contrib/awsxray/AwsMetricAttributeGenerator.java
Comment (mxiamxia) [reply aws-xray/src/main/java/io/opentelemetry/contrib/awsxray/AwsMetricAttributeGenerator.java:268]: changed it to: only add `http.method` as prefix for RemoteOperation when `http.target` is not null.

File: aws-xray/src/main/java/io/opentelemetry/contrib/awsxray/AwsMetricAttributeGenerator.java
Comment (mxiamxia) [reply aws-xray/src/main/java/io/opentelemetry/contrib/awsxray/AwsMetricAttributeGenerator.java:214]: it could happen in the following blocks which is not restricted by pair of values.

```
// try to derive RemoteService and RemoteOperation from the other related attributes
    if (remoteService.equals(UNKNOWN_REMOTE_SERVICE)) {
      remoteService = generateRemoteService(span);
    }
    if (remoteOperation.equals(UNKNOWN_REMOTE_OPERATION)) {
      remoteOperation = generateRemoteOperation(span);
    }
```

File: aws-xray/src/main/java/io/opentelemetry/contrib/awsxray/AwsAttributeKeys.java
Comment (mxiamxia) [reply aws-xray/src/main/java/io/opentelemetry/contrib/awsxray/AwsAttributeKeys.java:30]: Create an issue for better way to manage AWS specific attributes https://github.com/open-telemetry/opentelemetry-java-instrumentation/issues/8710

**Review by carolabadeer — APPROVED** [review]

File: aws-xray/src/main/java/io/opentelemetry/contrib/awsxray/AwsMetricAttributeGenerator.java
Comment (thpierce) [reply aws-xray/src/main/java/io/opentelemetry/contrib/awsxray/AwsMetricAttributeGenerator.java:217]: Discussed offline - not exactly what I was talking about, but we decided this is desirable behaviour.

**Review by thpierce — APPROVED** [review]

**Comment by mxiamxia** [comment]
Hi @trask, kindly follow up if you can help to review this PR and it has been reviewed by AWS reviewers. Thanks!

**Pull request merged**
